src/usr.bin/uuencode/uuencode.1
.\"	$MirOS: src/usr.bin/uuencode/uuencode.1,v 1.2 2005/03/13 18:33:52 tg Exp $
.\"	$OpenBSD: uuencode.1,v 1.16 2005/01/11 18:30:28 jmc Exp $
.\"	$FreeBSD: uuencode.1,v 1.26 2003/03/18 14:24:47 fanf Exp $
.\"
.\" Copyright (c) 1980, 1990, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)uuencode.1	8.1 (Berkeley) 6/6/93
.\" $FreeBSD$
.\"
.Dd January 27, 2002
.Dt UUENCODE 1
.Os
.Sh NAME
.Nm uuencode ,
.Nm uudecode ,
.Nm b64encode ,
.Nm b64decode
.Nd encode/decode a binary file
.Sh SYNOPSIS
.Nm uuencode
.Op Fl mr
.Op Fl o Ar output_file
.Op Ar file
.Ar name
.Nm uudecode
.Op Fl cimprs
.Op Ar
.Nm uudecode
.Op Fl i
.Fl o Ar output_file
.Nm b64encode
.Op Fl r
.Op Fl o Ar output_file
.Op Ar file
.Ar name
.Nm b64decode
.Op Fl ciprs
.Op Ar
.Nm b64decode
.Op Fl i
.Fl o Ar output_file
.Op Ar file
.Sh DESCRIPTION
The
.Nm uuencode
and
.Nm uudecode
utilities are used to transmit binary files over transmission mediums
that do not support formats other than printable
.Tn ASCII
data.
.Nm b64encode
and
.Nm b64decode
are equivalent to running
.Nm uuencode
and
.Nm uudecode
respectively with the
.Fl m
flag specified.
.Pp
.Nm uuencode
reads
.Ar file
(or by default, the standard input) and writes an encoded version
to the standard output, or to
.Ar output_file
if it has been specified.
The encoding uses only printing
.Tn ASCII
characters and includes the
mode of the file and the operand
.Ar name
for use by
.Nm uudecode .
.Pp
.Nm uudecode
transforms
.Dq uuencoded
files (or by default, the standard input) into the original form.
The resulting file is named either
.Ar name
or (depending on options passed to
.Nm uudecode )
.Ar output_file
and will have the mode of the original file except that set-user-ID
and execute bits are not retained.
.Nm uudecode
ignores any leading and trailing lines.
.Pp
The options for
.Nm uuencode
are as follows:
.Bl -tag -width ident
.It Fl m
Use the Base64 method of encoding, rather than the traditional
.Nm
algorithm.
.It Fl r
Do not write any framing lines.
This is a
.Mx
extension.
.It Fl o Ar output_file
Output to
.Ar output_file
instead of standard output.
.El
.Pp
The options for
.Nm uudecode
are as follows:
.Bl -tag -width ident
.It Fl c
Decode more than one uuencoded file from
.Ar file
if possible.
.It Fl i
Do not overwrite files.
.It Fl m
When used with the
.Fl r
flag, decode Base64 input instead of traditional
.Nm
input.
Without
.Fl r
it has no effect.
.It Fl o Ar output_file
Output to
.Ar output_file
instead of any pathname contained in the input data.
.It Fl p
Decode
.Ar file
and write output to standard output.
.It Fl r
Decode raw (or broken) input which is missing the initial and
possibly the final framing lines.
The input is assumed to be in the traditional
.Nm
encoding, but if the
.Fl m
flag is used, or if the utility is invoked as
.Nm b64decode ,
then the input is assumed to be in Base64 format.
.It Fl s
Do not strip output pathname to base filename.
By default
.Nm uudecode
deletes any prefix ending with the last slash '/' for security
reasons.
.El
.Sh EXAMPLES
The following example packages up a source tree, compresses it,
uuencodes it and mails it to a user on another system.
When
.Nm uudecode
is run on the target system, the file
.Pa src_tree.tar.Z
will be created which may then be uncompressed and extracted
into the original tree.
.Bd -literal -offset indent
$ tar cf - src_tree | compress | \e
uuencode src_tree.tar.Z | mail user@example.com
.Ed
.Pp
The following example unpacks all uuencoded
files from your mailbox into your current working directory.
.Bd -literal -offset indent
$ uudecode -c \*(Lt $MAIL
.Ed
.Pp
The following example extracts a compressed tar
archive from your mailbox
.Bd -literal -offset indent
$ uudecode -o /dev/stdout \*(Lt $MAIL | zcat | tar xfv -
.Ed
.Sh SEE ALSO
.Xr basename 1 ,
.Xr compress 1 ,
.Xr mail 1 ,
.Xr uuencode 5
.Sh STANDARDS
The
.Nm uudecode
and
.Nm uuencode
utilities conform to
.St -p1003.2-92 .
.Sh HISTORY
The
.Nm uudecode
and
.Nm
utilities appeared in
.Bx 4.0 .
.Pp
The
.Nm b64decode
and
.Nm b64encode
utilities appeared in
.Ox 3.6 ,
.Mx 8 .
.Sh BUGS
Files encoded using the traditional algorithm are expanded by 35%
(3 bytes become 4 plus control information).


src/usr.bin/uuencode/uuencode.c
/**	$MirOS$ */
/*	$OpenBSD: uuencode.c,v 1.7 2004/04/09 22:54:02 millert Exp $	*/
/*	$FreeBSD: uuencode.c,v 1.18 2004/01/22 07:23:35 grehan Exp $	*/

/*-
 * Copyright (c) 1983, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef lint
static const char copyright[] =
"@(#) Copyright (c) 1983, 1993\n\
	The Regents of the University of California.  All rights reserved.\n";
#endif /* not lint */

/*
 * Encode a file so it can be mailed to a remote system.
 */

#include <sys/param.h>
#include <sys/socket.h>
#include <sys/stat.h>

#include <netinet/in.h>

#include <err.h>
#include <locale.h>
#include <resolv.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

__SCCSID("@(#)uuencode.c	8.2 (Berkeley) 4/2/94");
__RCSID("$MirOS$");

void encode(void);
void base64_encode(void);
static void usage(void);

FILE *output;
int mode, rflag = 0;
char **av;

int
main(int argc, char *argv[])
{
	struct stat sb;
	int base64;
	int ch;
	char *outfile;
	extern char *__progname;

	base64 = 0;
	outfile = NULL;

	if (strcmp(__progname, "b64encode") == 0)
		base64 = 1;

	setlocale(LC_ALL, "");
	while ((ch = getopt(argc, argv, "mro:")) != -1) {
		switch (ch) {
		case 'm':
			base64 = 1;
			break;
		case 'r':
			rflag = 1;
			break;
		case 'o':
			outfile = optarg;
			break;
		case '?':
		default:
			usage();
		}
	}
	argv += optind;
	argc -= optind;

	switch(argc) {
	case 2:			/* optional first argument is input file */
		if (!freopen(*argv, "r", stdin) || fstat(fileno(stdin), &sb))
			err(1, "%s", *argv);
#define	RWX	(S_IRWXU|S_IRWXG|S_IRWXO)
		mode = sb.st_mode & RWX;
		++argv;
		break;
	case 1:
#define	RW	(S_IRUSR|S_IWUSR|S_IRGRP|S_IWGRP|S_IROTH|S_IWOTH)
		mode = RW & ~umask(RW);
		break;
	case 0:
		if (rflag)
			break;
		/* FALLTHROUGH */
	default:
		usage();
	}

	av = argv;

	if (outfile != NULL) {
		output = fopen(outfile, "w+");
		if (output == NULL)
			err(1, "unable to open %s for output", outfile);
	} else
		output = stdout;
	if (base64)
		base64_encode();
	else
		encode();
	if (ferror(output))
		errx(1, "write error");
	exit(0);
}

/* ENC is the basic 1 character encoding function to make a char printing */
#define	ENC(c) ((c) ? ((c) & 077) + ' ': '`')

/*
 * Copy from in to out, encoding in base64 as you go along.
 */
void
base64_encode(void)
{
	/*
	 * Output must fit into 80 columns, chunks come in 4, leave 1.
	 */
#define	GROUPS	((80 / 4) - 1)
	unsigned char buf[3];
	char buf2[sizeof(buf) * 2 + 1];
	size_t n;
	int rv, sequence;

	sequence = 0;

	if (!rflag)
		fprintf(output, "begin-base64 %o %s\n", mode, *av);
	while ((n = fread(buf, 1, sizeof(buf), stdin))) {
		++sequence;
		rv = b64_ntop(buf, n, buf2, (sizeof(buf2) / sizeof(buf2[0])));
		if (rv == -1)
			errx(1, "b64_ntop: error encoding base64");
		fprintf(output, "%s%s", buf2, (sequence % GROUPS) ? "" : "\n");
	}
	if (sequence % GROUPS)
		fprintf(output, "\n");
	if (!rflag)
		fprintf(output, "====\n");
}

/*
 * Copy from in to out, encoding as you go along.
 */
void
encode(void)
{
	int ch, n;
	char *p;
	char buf[80];

	if (!rflag)
		(void)fprintf(output, "begin %o %s\n", mode, *av);
	while ((n = fread(buf, 1, 45, stdin))) {
		ch = ENC(n);
		if (fputc(ch, output) == EOF)
			break;
		for (p = buf; n > 0; n -= 3, p += 3) {
			/* Pad with nulls if not a multiple of 3. */
			if (n < 3) {
				p[2] = '\0';
				if (n < 2)
					p[1] = '\0';
			}
			ch = *p >> 2;
			ch = ENC(ch);
			if (fputc(ch, output) == EOF)
				break;
			ch = ((*p << 4) & 060) | ((p[1] >> 4) & 017);
			ch = ENC(ch);
			if (fputc(ch, output) == EOF)
				break;
			ch = ((p[1] << 2) & 074) | ((p[2] >> 6) & 03);
			ch = ENC(ch);
			if (fputc(ch, output) == EOF)
				break;
			ch = p[2] & 077;
			ch = ENC(ch);
			if (fputc(ch, output) == EOF)
				break;
		}
		if (fputc('\n', output) == EOF)
			break;
	}
	if (ferror(stdin))
		errx(1, "read error");
	if (!rflag)
		(void)fprintf(output, "%c\nend\n", ENC('\0'));
}

static void
usage(void)
{
	(void)fprintf(stderr,
	    "usage: uuencode [-mr] [-o outfile] [infile] remotefile\n"
	    "       b64encode [-r] [-o outfile] [infile] remotefile\n");
	exit(1);
}
